### DISCUSSION TITLE
mingw64 memory leak tools

### DISCUSSION THREAD

**Discussion by yuansunliang1988** [jrfonseca/drmingw#90] [General]
Are there any good memory leak detection tools on mingw64? I've tried Dr.memory, and found that on win10 10 21H2 it gives an error, and the compiled debug version is not accurate to the line number!

**Comment by jrfonseca** [comment]
Dr. Memory was the only _"no recompilation needed"_ tool I knew for Windows -- a bit like Linux's Valgrind.  It's a pity it's not working.  https://github.com/DynamoRIO/drmemory appears to be maintained, so perhaps it's something that can be fixed.

For the record, https://github.com/jrfonseca/memtrail?tab=readme-ov-file#links I list many memory debugging tools, but all for Linux.

If you use  [Clang](https://github.com/mstorsjo/llvm-mingw) instead of GCC, then you can also use [AddressSanitizer](https://github.com/google/sanitizers/wiki/AddressSanitizer) to detect many kind of errors, but AFAICT,  [leak detection is not yet supported on Windows](https://github.com/google/sanitizers/wiki/AddressSanitizerLeakSanitizer).

I've successfully used https://github.com/johnsonjh/duma to debug memory leaks and buffer overflows with MinGW.   It was [a long time ago](https://gitlab.freedesktop.org/jrfonseca/mesa/-/commit/25ca815657514f6a21401f91a15c9c971a5c0e9e) but it worked great back then.  It's the only leak detection tool I can recommend for MinGW at this moment.

Good luck.  Please let me know in this issue if you had success with DUMA or any other tool.

**Reply by yuansunliang1988** [reply]
I'll try that, thanks for the reply, there really is so little support for ecology in this part of mingw.

**Reply by yuansunliang1988** [reply]
mingw64 Even with clang support, there is no memory leak detection, only memory overflow detection, DUMA This has a binary? I don't think I've seen it, only the source code is in the release.

**Reply by yuansunliang1988** [reply]
I don't see any mention of mingw or Windows in the list of things it supports.

**Reply by jrfonseca** [reply]
> mingw64 Even with clang support, there is no memory leak detection, only memory overflow detection

Yes, that's what I said.

> DUMA This has a binary? I don't think I've seen it, only the source code is in the release.

Not sure.  I just bundled the source code into my project when I used it with MinGW (refer to commit I linked to if you're curious.)

> I don't see any mention of mingw or Windows in the list of things it supports.

!? MinGW _is_ mentioned on https://github.com/johnsonjh/duma?tab=readme-ov-file#gnu-make

**Reply by yuansunliang1988** [reply]
It's finally done. I compiled the program in mingw's clang and used windbg's umdh to get it done, the new window_sdk install of windbg has a pothole and keeps reporting errors!
https://stackoverflow.com/questions/76383611/unable-to-use-umdh-from-windows-sdk-anymore
